### PR TITLE
Remove Orphan lock file

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
This PR make sure that CNI process is not blocked bacause of orphaned lock file. The cni will remove lock file if the process that is holding lockfile exited

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: cni/plugin.go [hunk 1]
```diff
@@ -198,3 +200,43 @@ func (plugin *Plugin) UninitializeKeyValueStore() error {
...
 	return nil
 }
+
+// check if safe to remove lockfile
+func (plugin *Plugin) IsSafeToRemoveLock(processName string) bool {
+	if plugin != nil && plugin.Store != nil {
+		// check if get process command supported
+		if cmdErr := platform.GetProcessSupport(); cmdErr != nil {
+			log.Errorf("Get process cmd not supported. Error %v", cmdErr)
+			return false
```
Comment (ashvindeodhar) [inline cni/plugin.go]: Should we return error along with bool? Here we are returning false if there is an error. In such cases if the lock file is left behind, we will never remove it and all the subsequent ADD calls will fail. Instead if we return the error, then we can create an ICM on it indicating that something is wrong and can look at the issue proactively before hitting the customer.

File: platform/os_windows.go [hunk 2]
```diff
@@ -60,6 +61,12 @@ func GetOSInfo() string {
 	return "windows"
 }
 
+func GetProcessSupport() error {
```
Comment (ashvindeodhar) [inline platform/os_windows.go]: same comment as above.

File: cni/plugin.go [hunk 3]
```diff
@@ -198,3 +200,43 @@ func (plugin *Plugin) UninitializeKeyValueStore() error {
...
+			log.Errorf("Num bytes read from lock file is 0")
+			return false
+		}
+
+		// Get the process name if running and
+		// check if that matches with our expected process
+		pid := strings.Trim(string(content), "\n")
+		pName, err := platform.GetProcessNameByID(pid)
+		if err != nil {
+			return true
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply cni/plugin.go]: the function GetProcessNameByID logs error incase of failure..so need to add here

File: cni/plugin.go [hunk 4]
```diff
@@ -198,3 +200,43 @@ func (plugin *Plugin) UninitializeKeyValueStore() error {
...
+		}
+
+		log.Printf("[CNI] Process name for pid %s is %s", pid, pName)
+
+		if pName != processName {
+			return true
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply cni/plugin.go]: i have already added log in other place before removing lock file

File: platform/os_linux.go [hunk 5]
```diff
@@ -44,6 +47,12 @@ func GetOSInfo() string {
 	return string(info)
 }
 
+func GetProcessSupport() error {
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply platform/os_linux.go]: This is to check if the cmd works as expected. If not, we should not rely on this function to remove lock file. In that case, its better we dont remove

File: cni/plugin.go (hunk 1 above)
Comment (tamilmani1989) [reply cni/plugin.go]: ok sure..let me return error also.

File: platform/os_linux.go [hunk 6]
```diff
@@ -127,3 +136,39 @@ func GetOSDetails() (map[string]string, error) {
 
 	return osInfoArr, nil
 }
+
+func IsProcessRunning(pidstr string) bool {
+	pid, err := strconv.Atoi(pidstr)
+	if err != nil {
+		log.Errorf("Cannot convert processid to int: %v", err)
+		return false
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply platform/os_linux.go]: sure

[the remark this answers is not in the document]
Comment (tamilmani1989) [reply platform/os_linux.go]: I think this function is not used..let me remove this

File: cni/ipam/plugin/main.go [hunk 7]
```diff
@@ -42,11 +42,19 @@ func main() {
 
 	if err := ipamPlugin.Plugin.InitializeKeyValueStore(&config); err != nil {
 		fmt.Printf("Failed to initialize key-value store of ipam plugin, err:%v.\n", err)
+
+		if isSafe, err := ipamPlugin.Plugin.IsSafeToRemoveLock(ipamPlugin.Plugin.Name); isSafe {
+			log.Printf("[IPAM] Removing lock file as process holding lock exited. Error:%v", err)
```
Comment (ashvindeodhar) [inline cni/ipam/plugin/main.go]: You are not logging the err when isSafe is false - which is where we get non-nil err.
You are always logging nil error.

File: cni/plugin.go [hunk 8]
```diff
@@ -198,3 +199,43 @@ func (plugin *Plugin) UninitializeKeyValueStore() error {
...
+			log.Errorf("Num bytes read from lock file is 0")
+			return false, fmt.Errorf("Num bytes read from lock file is 0")
+		}
+
+		log.Printf("Read from Lock file:%s", content)
+		// Get the process name if running and
+		// check if that matches with our expected process
+		pName, err := platform.GetProcessNameByID(string(content))
+		if err != nil {
+			return true, nil
```
Comment (ashvindeodhar) [inline cni/plugin.go]: if GetProcessNameByID returned err due to some internal err, we would end up deleting the lock file even if the process which acquired the lock is running.
I think we should return false here. Let's discuss this if you want.

File: platform/os_linux.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashvindeodhar) [reply platform/os_linux.go]: What I am saying is, this is redundant function / call. You are doing exactly the same thing as a part of GetProcessNameByID. You can just return saying don't remove if that returns any error.

**Pull request merged**